examples/cis2-wccd-upgradable/src/lib.rs

```diff
@@ -0,0 +1,3537 @@
+                from:     sender,
+                data:     input.params.data,
+            };
+            host.invoke_contract(
+                &address,
+                &parameter,
+                function.as_entrypoint_name(),
+                Amount::zero(),
+            )
+            .unwrap_abort();
+        }
+    }
+    Ok(())
+}
+
+/// Unwrap an amount of wCCD tokens into CCD
+#[receive(contract = "CIS2-wCCD", name = "unwrap", mutable)]
+fn contract_unwrap<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<StateImplementation, StateApiType = S>,
+) -> ContractResult<()> {
+    let (proxy_address, state_address) = get_protocol_addresses_implementation(host)?;
+
+    // Can be only called through the fallback function on the proxy.
+    only_proxy(proxy_address, ctx.sender())?;
+
+    // Check that contract is not paused.
+    when_not_paused(&state_address, host)?;
+
+    let input: UnwrapParamsWithSender = ctx.parameter_cursor().get()?;
+
+    if input.params.amount == 0u64.into() {
+        return Ok(());
+    }
+
+    // Get the sender who invoked this contract function.
+    let sender = input.sender;
+
+    ensure!(
+        sender == input.params.owner
+            || host
+                .state()
+                .is_operator(&state_address, &sender, &input.params.owner, host)
+                .map_or(false, |is_operator| is_operator),
+        ContractError::Unauthorized
+    );
+
+    // Update the state.
+
+    let token_balance =
+        host.state().balance(&state_address, &TOKEN_ID_WCCD, &input.params.owner, host)?;
+
+    ensure!(token_balance >= input.params.amount, ContractError::InsufficientFunds);
+
+    let set_balance_params = SetBalanceParams {
+        owner:  input.params.owner,
+        amount: input.params.amount,
+        update: Update::Remove,
+    };
+
+    let parameter_bytes = to_bytes(&set_balance_params);
+
+    host.invoke_contract_raw(
+        &state_address,
+        Parameter(&parameter_bytes),
+        EntrypointName::new_unchecked("setBalance"),
+        Amount::zero(),
+    )?;
+
+    let params2: TransferCCDParams = TransferCCDParams {
+        amount: input.params.amount,
+    };
+
+    let parameter_bytes = to_bytes(&params2);
+
+    // Access/Request CCD funds from proxy.
+    host.invoke_contract_raw(
+        &proxy_address,
+        Parameter(&parameter_bytes),
+        EntrypointName::new_unchecked("transferCCD"),
+        Amount::zero(),
+    )?;
+
+    let unwrapped_amount = Amount::from_micro_ccd(input.params.amount.into());
+
+    // Send CCD to receiver.
+    match input.params.receiver {
+        Receiver::Account(address) => host.invoke_transfer(&address, unwrapped_amount)?,
+        Receiver::Contract(address, function) => {
+            host.invoke_contract(
+                &address,
+                &input.params.data,
+                function.as_entrypoint_name(),
+                unwrapped_amount,
+            )?;
+        }
+    }
```

Comment:
Just a thought:
If the `transferCCD` function took a receiver in the parameter, it could send the CCD directly instead of the implementation having to do it.
The main benefit is that users will receive the unwrapped CCD from the Proxy contracts instead of the implementation contract.

Reply:
The proxy transfers the CCD now directly to the receiver address.